importer/mcp/fetcher.go

```diff
@@ -0,0 +1,196 @@
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"slices"
+	"strconv"
+	"time"
+
+	mcpapiv0 "github.com/modelcontextprotocol/registry/pkg/api/v0"
+)
+
+const (
+	// defaultPageLimit is the default number of servers to fetch per page.
+	defaultPageLimit = 30
+)
+
+// Supported filters https://registry.modelcontextprotocol.io/docs#/operations/list-servers#Query-Parameters
+//   - search: Filter by server name (substring match)
+//   - version: Filter by version ('latest' for latest version, or an exact version like '1.2.3')
+//   - updated_since: Filter by updated time (RFC3339 datetime)
+//   - limit: Number of servers per page (default 30)
+//   - cursor: Pagination cursor
+var supportedFilters = []string{
+	"search",
+	"version",
+	"updated_since",
+	"limit",
+	"cursor",
+}
+
+// Fetcher implements the pipeline.Fetcher interface for MCP registry.
+type Fetcher struct {
+	baseURL    string
+	httpClient *http.Client
+	filters    map[string]string
+	limit      int
+}
+
+// NewFetcher creates a new MCP fetcher.
+func NewFetcher(baseURL string, filters map[string]string, limit int) *Fetcher {
+	return &Fetcher{
+		baseURL: baseURL,
+		httpClient: &http.Client{
+			Timeout: 30 * time.Second, //nolint:mnd
+		},
+		filters: filters,
+		limit:   limit,
+	}
+}
+
+// Fetch retrieves servers from the MCP registry and sends them to the output channel.
+func (f *Fetcher) Fetch(ctx context.Context) (<-chan interface{}, <-chan error) {
+	outputCh := make(chan interface{})
+	errCh := make(chan error, 1)
+
+	// Validate filters
+	for key := range f.filters {
+		if !slices.Contains(supportedFilters, key) {
+			close(outputCh)
+
+			errCh <- fmt.Errorf("unsupported filter: %s", key)
+
+			close(errCh)
+
+			return outputCh, errCh
+		}
+	}
+
+	go func() {
+		defer close(outputCh)
+		defer close(errCh)
+
+		cursor := ""
+		count := 0
+
+		for {
+			// Check context cancellation
+			select {
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+
+				return
+			default:
+			}
+
+			// Fetch one page
+			page, nextCursor, err := f.listServersPage(ctx, cursor)
+			if err != nil {
+				errCh <- err
+
+				return
+			}
+
+			// Stream each server as soon as it's available
+			for _, server := range page {
+				// Check if limit is reached (limit <= 0 means no limit)
+				if f.limit > 0 && count >= f.limit {
+					return
+				}
+
+				select {
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+
+					return
+				case outputCh <- server:
+					count++
+				}
+			}
+
+			// Check if there are more pages
+			if nextCursor == "" {
+				break
+			}
+
+			cursor = nextCursor
+		}
+	}()
+
+	return outputCh, errCh
+}
+
+// listServersPage fetches a single page of servers from the MCP registry.
+func (f *Fetcher) listServersPage(ctx context.Context, cursor string) ([]mcpapiv0.ServerResponse, string, error) {
+	// Build URL with query parameters
+	u, err := url.Parse(f.baseURL + "/servers")
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to parse base URL: %w", err)
+	}
```

Comment:
Move this to the constructor function and store `baseURL` as `*url.URL` instead of `string`. No need to parse it every call.